riscv-page-tables/src/page_table.rs

```diff
@@ -585,6 +600,72 @@ impl<T: PagingMode> PageTableInner<T> {
         entry
     }
 
+    /// Creates contiguous mappings, both physically and virtually, of `num_pages` pages, starting
+    /// from `start_vaddr` and `start_paddr`, with the same permissions.
+    ///
+    /// # Safety
+    ///
+    /// The caller must guarantee that app pages reference from `start_paddr` for `num_pages` are
+    /// uniquely owned by the root `GuestStagePageTable`.
+    #[allow(dead_code)]
+    pub unsafe fn map_contiguous_leaves(
+        &mut self,
+        start_vaddr: PageAddr<T::MappedAddressSpace>,
+        start_paddr: SupervisorPageAddr,
+        num_pages: u64,
+        page_size: PageSize,
+        perms: PteFieldBits,
+    ) -> Result<()> {
+        start_vaddr
+            .checked_add_pages_with_size(num_pages, page_size)
+            .ok_or(Error::OutOfMapRange)?;
+        start_paddr
+            .checked_add_pages_with_size(num_pages, page_size)
+            .ok_or(Error::OutOfMapRange)?;
+
+        let mut remaining = num_pages;
+        let mut vaddr = start_vaddr;
+        let mut paddr = start_paddr;
+
+        while remaining > 0 {
+            let mut table = PageTable::from_root(self);
+            // Get the page-table at requested level.
+            while table.level.leaf_page_size() != page_size {
+                table = table.next_level(RawAddr::from(vaddr))?;
+            }
+
+            let start_idx = table.index_from_addr(vaddr.raw());
+            let mut done = 0;
+            for i in start_idx.iter().take(remaining as usize) {
+                let entry = table.entry_for_index_mut(i);
+                use TableEntryType::*;
+                match entry {
+                    LockedUnmapped(l) => {
+                        if l.level().leaf_page_size() != page_size {
+                            return Err(Error::PageSizeMismatch(
+                                page_size,
+                                l.level().leaf_page_size(),
+                            ));
+                        }
```

Comment:
don't we know that the page size here matches based on walking the page tables in the loop above?

Reply:
Yes, I was just copying what was in map_leaf because this code seems temporary anyway. If we support different page sizes then table should not be unreachable there. So I kind of wanted to understand what the status is, and in general not interfere with future works on huge pages.